codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/auth/AuthSchemeResolverGenerator.kt

```diff
@@ -0,0 +1,237 @@
+                        }
+                    }
+                }
+                rust(".into()")
+            }
+        }
+
+    /**
+     * Return [RuntimeType] representing the per-service trait definition for auth scheme resolution
+     */
+    fun serviceSpecificResolveAuthSchemeTrait(): RuntimeType {
+        return RuntimeType.forInlineFun("ResolveAuthScheme", ClientRustModule.Config.auth) {
+            rustTemplate(
+                """
+                /// Auth scheme resolver trait specific to ${codegenContext.serviceShape.serviceNameOrDefault("this service")}
+                pub trait ResolveAuthScheme: #{Send} + #{Sync} + #{Debug} {
+                    /// Resolve a priority list of auth scheme options with the given parameters
+                    fn resolve_auth_scheme<'a>(
+                        &'a self,
+                        params: &'a #{Params},
+                        cfg: &'a #{ConfigBag},
+                        runtime_components: &'a #{RuntimeComponents},
+                    ) -> #{AuthSchemeOptionsFuture}<'a>;
+
+                    /// Convert this service-specific resolver into a `SharedAuthSchemeOptionResolver`
+                    fn into_shared_resolver(self) -> #{SharedAuthSchemeOptionResolver}
+                    where
+                        Self: #{Sized} + 'static,
+                    {
+                        #{SharedAuthSchemeOptionResolver}::new(DowncastParams(self))
+                    }
+                }
+
+                ##[derive(Debug)]
+                struct DowncastParams<T>(T);
+                impl<T> #{ResolveAuthSchemeOptions} for DowncastParams<T>
+                where
+                    T: ResolveAuthScheme,
+                {
+                    fn resolve_auth_scheme_options_v2<'a>(
+                        &'a self,
+                        params: &'a #{AuthSchemeOptionResolverParams},
+                        cfg: &'a #{ConfigBag},
+                        runtime_components: &'a #{RuntimeComponents},
+                    ) -> #{AuthSchemeOptionsFuture}<'a> {
+                        match params.get::<#{Params}>() {
+                            #{Some}(concrete_params) => self.0.resolve_auth_scheme(concrete_params, cfg, runtime_components),
+                            #{None} => #{AuthSchemeOptionsFuture}::ready(#{Err}("params of expected type was not present".into())),
+                        }
+                    }
+                }
+                """,
+                *codegenScope,
+            )
+        }
+    }
+
+    // TODO(https://github.com/smithy-lang/smithy-rs/issues/4177): Only used in protocol tests.
+    //  Remove this once the issue has been addressed.
+    internal fun noAuthSchemeResolver(): RuntimeType {
+        return RuntimeType.forInlineFun("NoAuthSchemeResolver", ClientRustModule.Config.auth) {
+            rustTemplate(
+                """
+                ##[allow(dead_code)] // since usage is behind `cfg(test)`
```

Comment:
If usage is only `cfg(test)` should we cfg gate this as well?

Reply:
Switched to `cfg(test)` completely in 84b717b.